src/renderer/atlas/AtlasEngine.r.cpp

```diff
@@ -119,23 +125,103 @@ catch (const wil::ResultException& exception)
+}
+
+void AtlasEngine::WaitUntilCanRender() noexcept
+{
+    // IDXGISwapChain2::GetFrameLatencyWaitableObject returns an auto-reset event.
+    // Once we've waited on the event, waiting on it again will block until the timeout elapses.
+    // _r.waitForPresentation guards against this.
+    if (!debugGeneralPerformance && std::exchange(_r.waitForPresentation, false))
+    {
+        WaitForSingleObjectEx(_r.frameLatencyWaitableObject.get(), 100, true);
+#ifndef NDEBUG
+        _r.frameLatencyWaitableObjectUsed = true;
+#endif
+    }
+}
+
 #pragma endregion
 
-void AtlasEngine::_setShaderResources() const
+void AtlasEngine::_renderWithCustomShader() const
 {
-    _r.deviceContext->VSSetShader(_r.vertexShader.get(), nullptr, 0);
-    _r.deviceContext->PSSetShader(_r.pixelShader.get(), nullptr, 0);
+    // Render with our main shader just like Present().
+    {
+        // OM: Output Merger
+        _r.deviceContext->OMSetRenderTargets(1, _r.customOffscreenTextureTargetView.addressof(), nullptr);
+        _r.deviceContext->Draw(3, 0);
+    }
+
+    // Update the custom shader's constant buffer.
+    {
+        CustomConstBuffer data;
+        data.time = std::chrono::duration<float>(std::chrono::steady_clock::now() - _r.customShaderStartTime).count();
+        data.scale = _r.pixelPerDIP;
+        data.resolution.x = static_cast<float>(_r.cellCount.x * _r.fontMetrics.cellSize.x);
+        data.resolution.y = static_cast<float>(_r.cellCount.y * _r.fontMetrics.cellSize.y);
+        data.background = colorFromU32<f32x4>(_r.backgroundColor);
+
+#pragma warning(suppress : 26494) // Variable 'mapped' is uninitialized. Always initialize an object (type.5).
+        D3D11_MAPPED_SUBRESOURCE mapped;
+        THROW_IF_FAILED(_r.deviceContext->Map(_r.customShaderConstantBuffer.get(), 0, D3D11_MAP_WRITE_DISCARD, 0, &mapped));
+        assert(mapped.RowPitch >= sizeof(data));
+        memcpy(mapped.pData, &data, sizeof(data));
+        _r.deviceContext->Unmap(_r.customShaderConstantBuffer.get(), 0);
+    }
+
+    // Render with the custom shader.
+    {
+        // OM: Output Merger
+        // customOffscreenTextureView was just rendered to via customOffscreenTextureTargetView and is
+        // set as the output target. Before we can use it as an input we have to remove it as an output.
+        _r.deviceContext->OMSetRenderTargets(1, _r.renderTargetView.addressof(), nullptr);
+
+        // VS: Vertex Shader
+        _r.deviceContext->VSSetShader(_r.customVertexShader.get(), nullptr, 0);
+
+        // PS: Pixel Shader
+        _r.deviceContext->PSSetShader(_r.customPixelShader.get(), nullptr, 0);
+        _r.deviceContext->PSSetConstantBuffers(0, 1, _r.customShaderConstantBuffer.addressof());
+        _r.deviceContext->PSSetShaderResources(0, 1, _r.customOffscreenTextureView.addressof());
+        _r.deviceContext->PSSetSamplers(0, 1, _r.customShaderSamplerState.addressof());
+
+        _r.deviceContext->Draw(4, 0);

```

Comment:
4 is a weird magic number; what is it?

Reply:
The 4 vertices form the 2 triangles covering the rectangular viewport.